src/little_set.jl

```diff
@@ -0,0 +1,345 @@
+
+struct LittleSet{T, D<:StoreType{T}} <: AbstractSet{T}
+    data::D
+
+    global const UnfrozenLittleSet{T} = LittleSet{T, <:AbstractVector{T}}
+    global const FrozenLittleSet{T} = LittleSet{T, <:Tuple}
+    global const OpaqueLittleSet{T} = LittleSet{T, Tuple{Vararg{T}}}
+
+    LittleSet{T, D}(data) where {T,D} = new{T, D}(data)
+    function OpaqueLittleSet{T}(@nospecialize(data)) where {T}
+        if isa(data, Tuple{Vararg{T}})
+            new{T, Tuple{Vararg{T}}}(data)
+        else
+            new{T, Tuple{Vararg{T}}}(Tuple(data))
+        end
+    end
+    function OpaqueLittleSet(@nospecialize(data))
+        T = eltype(data)
+        new{T, Tuple{Vararg{T}}}(data)
+    end
+    function LittleSet{T}(data::AbstractVector) where {T}
+        if eltype(data) == T
+            return new{T, typeof(data)}(data)
+        else
+            d = convert(AbstractVector{T}, data)
+            return new{T, typeof(d)}(d)
+        end
+    end
+    LittleSet{T}(data::Tuple{Vararg{T}}) where {T} = new{T, typeof(data)}(data)
+    LittleSet{T}(itr) where {T} = union!(LittleSet{T}(), itr)
+    LittleSet{T}() where {T} = new{T, Vector{T}}(Vector{T}())
+    LittleSet() = LittleSet{Any}()
+    function LittleSet(itr)
+        IET = Base.IteratorEltype(itr)
+        if isa(IET, Base.HasEltype)
+            LittleSet{eltype(itr)}(itr)
+        else
+            T = Base.@default_eltype(itr)
+            if (Base.isconcretetype(T) || T === Union{})
+                return LittleSet{T}(itr)
+            else
+                return Base.grow_to!(LittleSet{T}(), itr)
+            end
+        end
+    end
+end
+
+# in cases where tuple parameters have been intentionally made opaque (such as
+# `FrozenLittleSet{T, Tuple{Vararg{T}}}`), these methods allow accessing data without exposing the
+# underlying `data` field and unintentionally specializing downstream code on the exact type
+# representation of a tuple.
+if isdefined(Base, Symbol("@assume_effects"))
```

Comment:
This doesn't work in Julia 1.6 (see CI failure). Perhaps this will?
```suggestion
@static if isdefined(Base, Symbol("@assume_effects"))
```